File: solutions/zig/collatz-conjecture/1/collatz_conjecture.zig
pub const ComputationError = error {
    IllegalArgument,
};

pub fn steps(number: usize) anyerror!usize {
    return switch (number) {
      0 => ComputationError.IllegalArgument,
      1 => 0,
      else => steps_acc(number, 0),
    };
}

fn steps_acc(n: usize, acc: usize) usize {
    return switch (n) {
      0 => 0,
      1 => acc,
      else => if (n % 2 == 0)
                steps_acc(n / 2, acc + 1)
              else 
                steps_acc(3 * n + 1, acc + 1),
    };
}

File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

static int steps_acc(int n, int acc)
{
    if (n == 1)
        return acc;
    else if (n % 2 == 0)
        return steps_acc(n / 2, acc + 1);
    else
        return steps_acc(n * 3 + 1, acc + 1);
}

int steps(int start)
{
    if (start <= 0)
        return ERROR_VALUE;
    else
        return steps_acc(start, 0);
}